fix(java): invalid collections returned with non-class elements

RomainMuller opened the pull request:
The Java runtime for jsii used to not be able to necessarily hold the
type information necessary to correctly process the contents of
collections, as the generic information is lost.

In order to address this, introduced a new `NativeType<T>` class that is
able to carry:
1. The actual declared type of the value (via generic type parameter)
2. The raw type of the value (reified)
3. The correct JavaType to pass to Jackson to obtain the correct result

It also preserves the specific behavior around de-serialization of
`JsiiObject` sub-classes as well as jsii interfaces, correctly
requesting de-serialization as the `JsiiObject` class, and in a second
phase `transform` this into the correct value type as needed.

Introduced a new test that validates the `List<T>` and `Map<String, T>`
values received from `node` are deserialized to instances with the
appropriate content type.

Fixes #1196

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

---

## Commit Message
fix(java): invalid collections returned with non-class elements (#1197)

The Java runtime for jsii used to not be able to necessarily hold the
type information necessary to correctly process the contents of
collections, as the generic information is lost.

In order to address this, introduced a new `NativeType<T>` class that is
able to carry:
1. The actual declared type of the value (via generic type parameter)
2. The raw type of the value (reified)
3. The correct JavaType to pass to Jackson to obtain the correct result

It also preserves the specific behavior around de-serialization of
`JsiiObject` sub-classes as well as jsii interfaces, correctly
requesting de-serialization as the `JsiiObject` class, and in a second
phase `transform` this into the correct value type as needed.

Introduced a new test that validates the `List<T>` and `Map<String, T>`
values received from `node` are deserialized to instances with the
appropriate content type.

Fixes #1196

RomainMuller commented:
Labelled with `do-not-merge` as I'm pondering whether the old code paths need to be preserved in order to ensure proper interoperability with code generated against older versions of the Java runtime.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dkba3h5q98.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=28bdb0b0-d194-4dd8-b07d-088be092da46%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://0j8j2mdtcg.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d820364c-2718-422a-ad07-190b2b2b08eb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

rix0rrr commented:
To help me review, since this change is ungodly:

This was flagged as a regression (since jsii 0.17 if I'm reading the breadcrumb trail directly). Can you explain succinctly what the changes were that caused this breakage (starting from the wire protocol changes, if any) and what the approach for the fix is?

To the uncultured plebeian that I am, it would seem that if the wire protocol says:

```
[ { "jsii$ref": "my.MyObject@1" }]
```

It should be easy enough to construct a `List` with a `MyObject` in it?

It's probably trickier than that but I need it explained to me like I'm 5 :).

RomainMuller commented:
I significantly reduced the amount of regression test file changes by preserving the previous API (and having it delegate into the new one); and only switching generated code to the new API when the older one was unsafe.

To answer @rix0rrr's question, here's what happened:
* In #825, the Object ID generation was changed, specifically around interfaces.
  + The format of Object IDs generally speaking is `<type-fqn>@<some-number>`.
  + Prior to #825, the `<type-fqn>` part could be an interface if the value was exiting JavaScript through an interface-typed method or property.
  + After #825, the `<type-fqn>` part is *required* to be a class name, or `Object` (pure interface implementations, and by-reference structs). Necessary interface declarations are passed on a separate field of the `ObjRef`.
* This means that since then, first pass hydration in Java cannot yield an interface at all... Obtaining an interface view of an object requires on additional step (`JsiiObject.asInterfaceProxy(Class<?>)`).
* This additional step is done for single values, but was missed from `List`s and `Map`s; because the only signal that is passed to the deserializer was `List.class` or `Map.class`, which yields no information about the expected element type.
* Add to this that Java does not allow providing customized implicit conversions (the ways C# does), nor creating a dynamic proxy object with a base class other than Object (the way it can be done in Python), and you get the broken behavior.

Hopefully this helps clarify the problem...

Now - to the approach for solving this: I introduced a way to represent complex types to provide directions to the deserializer, in the form of the `NativeType` class. This is a Jackson (our JSON handling tool of choice) friendly way to express the expected type to be deserialized, including any specializations such as element type.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://0j8j2mdtcg.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7126807b-e4dd-417a-a556-27cf8947ec4a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dkba3h5q98.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=49e4565d-10df-4027-9bca-5f8002b8a430%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

eladb commented:
I am not sure I fully understand if this is a breaking change. I don't think we can afford breaking changes neither of the manifest nor of the runtime protocol.

RomainMuller commented:
@eladb - I don't think this changes the runtime protocol. The only part that may be "broken" here is technically not working correctly in the current state of things. The only one thing which I think may still need "easing in" (so new runtime works with code generated by the previous generator) is with how the "old API" was modified to delegate into the "new API", which will throw if one attempts to use an api that returns a List or a Map (those could have returned collection with "illegal" items in them).

I think I can work out a way to make the old behavior preserved, so older generated code keeps behaving the same as it currently does (still broken when it was, but not failing more).

I can then attempt to run a number of scenarios using the current CDK libraries with the new Java runtime ahead of the class path to validate it is indeed backwards-compatible?

I'd add unit tests to assert persistence of the older behavior, but I'd be skeptical as to the coverage of such (new) tests being sufficient...

rix0rrr commented:
> * This means that since then, first pass hydration in Java cannot yield an interface at all...

This part I don't get.  It used to be:

```
{ "jsii$ref": "my.MyInterface@1" }
```

And now it's:

```
{ "jsii$ref": "Object@1", "jsii$interfaces": ["my.MyInterface"] }
```

We can restore the "old" behavior by picking the first interface from the associated list and instantiating the proxy class for that type, right?

I guess we're doing all this work so that if there are 2 interfaces in the wire list, we'll pick the one that the current  method has been declared to return? (Without looking at the wire data, looks like, just assuming the types will work out as declared?)

I am *slightly* concerned about all the introspection happening on every JSII call; every call constructs a set of new `NativeType` objects, which could be cached because they're always the same. But I guess this is not the most expensive work happening on jsii calls :)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://0j8j2mdtcg.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f074390b-d763-4d65-b077-c1045113af39%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dkba3h5q98.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e5279952-f6c8-4c38-a5c7-c7879af1236b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dkba3h5q98.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fab0bd37-a899-45c7-9d62-8962bbac64b3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://0j8j2mdtcg.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=caa21a5a-3927-41f0-b330-1943c6fae665%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

RomainMuller commented:
I added a test of the *previous* behavior with respects to `List` and `Map`, in order to ensure the new code path does not cause older generated code to crash. I also fixed the new code path so that... it... does not crash on such code paths.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dkba3h5q98.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6b44b078-3a57-4395-843f-9e22342a142b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://0j8j2mdtcg.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ff10a18c-f88c-4c74-b83d-c677496469b2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

eladb commented on packages/@jsii/java-runtime/pom.xml.t.js:
```diff
@@ -113,10 +113,10 @@ process.stdout.write(`<?xml version="1.0" encoding="UTF-8"?>
         </dependency>
 
-        <!-- https://mvnrepository.com/artifact/org.mockito/mockito-core -->
+        <!-- https://mvnrepository.com/artifact/org.mockito/mockito-junit-jupiter -->
```
Why was this changed?

eladb commented on packages/@jsii/java-runtime/project/src/main/java/software/amazon/jsii/NativeType.java:
```diff
@@ -0,0 +1,216 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public abstract class NativeType<T> {
```
Missing documentation

eladb commented on packages/@jsii/java-runtime/project/src/main/java/software/amazon/jsii/NativeType.java:
```diff
@@ -0,0 +1,216 @@
+import java.util.stream.Collectors;
+
+public abstract class NativeType<T> {
+    protected static final JavaType[] NO_TYPE_PARAMS = {};
```
Missing docs
